### **Bài 1. Tạo Tensor cơ bản**
Tạo một Tensor số nguyên kích thước 2 × 3 và hiển thị kết quả.

In [1]:
import torch
a = torch.tensor(
[[1, 2, 3],
[4, 5, 6]],
dtype=torch.int32
)
print(a)

tensor([[1, 2, 3],
        [4, 5, 6]], dtype=torch.int32)


Kích thước: (2, 3)

Số chiều: 2

Kiểu dữ liệu: torch.int32

### **Bài 2. Tạo các Tensor thường dùng**
Thực hành tạo Tensor ngẫu nhiên, Tensor toàn số 0, toàn số 1 và ma trận đơn vị.

In [2]:
import torch
torch.manual_seed(42)

a = torch.rand(3, 4)
b = torch.randn(3, 4)
c = torch.zeros(2, 3)
d = torch.ones(2, 3)
e = torch.eye(4)
f = torch.full((5, 5), 10)

print(a)
print(b)
print(c)
print(d)
print(e)
print(f)

tensor([[0.8823, 0.9150, 0.3829, 0.9593],
        [0.3904, 0.6009, 0.2566, 0.7936],
        [0.9408, 0.1332, 0.9346, 0.5936]])
tensor([[ 2.2082, -0.6380,  0.4617,  0.2674],
        [ 0.5349,  0.8094,  1.1103, -1.6898],
        [-0.9890,  0.9580,  1.3221,  0.8172]])
tensor([[0., 0., 0.],
        [0., 0., 0.]])
tensor([[1., 1., 1.],
        [1., 1., 1.]])
tensor([[1., 0., 0., 0.],
        [0., 1., 0., 0.],
        [0., 0., 1., 0.],
        [0., 0., 0., 1.]])
tensor([[10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10],
        [10, 10, 10, 10, 10]])


a (torch.rand): Tensor kích thước (3, 4) với các giá trị ngẫu nhiên trong khoảng [0, 1).

b (torch.randn): Tensor kích thước (3, 4) với các giá trị ngẫu nhiên theo phân phối chuẩn (mean=0, variance=1).

c (torch.zeros): Tensor kích thước (2, 3) toàn số 0.

d (torch.ones): Tensor kích thước (2, 3) toàn số 1.

e (torch.eye): Ma trận vuông đơn vị kích thước (4, 4) (đường chéo bằng 1, còn lại bằng 0).

f (torch.full): Tensor kích thước (5, 5) với mọi phần tử bằng 10, tạo bằng `torch.full((5, 5), 10)`.

### **Bài 3. Kiểm tra thông tin Tensor**
Sử dụng shape, size(), ndim và dtype để kiểm tra các thuộc tính của Tensor.

In [3]:
import torch
a = torch.zeros(2, 3, 4)
print("Tensor:")
print(a)
print("Shape:", a.shape)
print("Size:", a.size())
print("Number of dimensions:", a.ndim)
print("Data type:", a.dtype)

Tensor:
tensor([[[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]],

        [[0., 0., 0., 0.],
         [0., 0., 0., 0.],
         [0., 0., 0., 0.]]])
Shape: torch.Size([2, 3, 4])
Size: torch.Size([2, 3, 4])
Number of dimensions: 3
Data type: torch.float32


torch.Size([2, 3, 4]) có nghĩa đây là một Tensor 3 chiều (3D), cấu trúc gồm:

- 2: Số lượng khối (hoặc số ma trận 2D).

- 3: Số hàng (row) trong mỗi ma trận.

- 4: Số cột (column) trong mỗi ma trận.

### **Bài 4. Slicing và Indexing**
Truy xuất một phần dữ liệu từ Tensor nhiều chiều.

In [4]:
import torch
a = torch.randn(3, 4, 5)
print("Original tensor:")
print(a)
print("\nTensor a[1]:")
print(a[1])
print("\nTensor a[1:, 2:4]:")
print(a[1:, 2:4])

Original tensor:
tensor([[[ 3.1888e-01, -4.2452e-01,  3.0572e-01, -7.7459e-01,  3.4912e-02],
         [ 3.2110e-01,  1.5736e+00, -8.4547e-01, -1.2742e+00,  2.1228e+00],
         [-1.2347e+00, -4.8791e-01, -1.4181e+00,  8.9627e-01,  4.9905e-02],
         [ 2.2667e+00, -4.8799e-01,  1.1914e+00, -8.1401e-01, -7.3599e-01]],

        [[-8.3712e-01, -9.2239e-01,  1.8113e+00,  1.6056e-01, -9.7807e-02],
         [ 1.8446e+00, -1.1845e+00,  1.3835e+00, -1.2024e+00,  7.0781e-01],
         [-1.0759e+00,  5.3565e-01,  3.4665e-01, -1.9733e-01, -1.0546e+00],
         [ 1.2780e+00,  1.4534e-01,  2.3105e-01,  8.6540e-03, -1.4229e-01]],

        [[ 5.7501e-01, -6.4172e-01, -2.2064e+00, -7.5080e-01, -2.5095e+00],
         [ 4.8800e-01,  7.8459e-01,  2.8647e-02,  6.4076e-01,  5.8325e-01],
         [ 1.0669e+00, -4.5015e-01, -1.8527e-01,  7.5276e-01,  4.0476e-01],
         [ 1.7847e-01,  2.6491e-01,  1.2732e+00, -1.3109e-03, -3.0360e-01]]])

Tensor a[1]:
tensor([[-0.8371, -0.9224,  1.8113,  0.1606, -0.097

Shape của a[1]: (4, 5)

Shape của a[1:, 2:4]: (2, 2, 5)

Lấy hai phần tử cuối cùng trên chiều thứ ba: Cú pháp a[:, :, -2:] (hoặc a[..., -2:]). Tensor kết quả có shape là (3, 4, 2).

In [5]:
a = torch.randn(3, 4, 5)
result = a[:, :, -2:]
print(a.shape)
print(result.shape)
print(result)

torch.Size([3, 4, 5])
torch.Size([3, 4, 2])
tensor([[[ 0.4771,  0.7262],
         [-0.0127,  0.2408],
         [ 0.3399,  0.7200],
         [-1.4364, -1.1299]],

        [[ 0.5760,  1.1415],
         [-0.3753,  1.0331],
         [ 0.9585,  1.6192],
         [-0.7328,  0.1043]],

        [[ 1.6048, -1.2956],
         [ 0.0358,  0.1206],
         [-0.9637, -0.3750],
         [-1.9267,  0.1279]]])


### **Bài 5. Reshape và Flatten**
Thay đổi hình dạng Tensor mà không làm thay đổi số phần tử.

In [6]:
import torch
a = torch.randn(3, 4)
print("Original:")
print(a)
print("\nFlatten:")
print(a.ravel())
print("\nReshape to 3 x 2 x 2:")
print(a.reshape(3, 2, 2))

Original:
tensor([[ 0.7764, -0.3029, -1.2753, -0.4758],
        [ 2.3839,  0.9157, -0.6430,  0.7113],
        [ 0.4000, -1.2039, -0.4198, -1.1929]])

Flatten:
tensor([ 0.7764, -0.3029, -1.2753, -0.4758,  2.3839,  0.9157, -0.6430,  0.7113,
         0.4000, -1.2039, -0.4198, -1.1929])

Reshape to 3 x 2 x 2:
tensor([[[ 0.7764, -0.3029],
         [-1.2753, -0.4758]],

        [[ 2.3839,  0.9157],
         [-0.6430,  0.7113]],

        [[ 0.4000, -1.2039],
         [-0.4198, -1.1929]]])


Số phần tử của Tensor trước và sau khi reshape luôn không đổi là 12 ($3 \times 4 = 12$ cho bản gốc và $3 \times 2 \times 2 = 12$ sau khi reshape)

reshape 2×6: a.reshape(2, 6)

In [7]:
a_2x6 = a.reshape(2, 6)
print(a_2x6)

tensor([[ 0.7764, -0.3029, -1.2753, -0.4758,  2.3839,  0.9157],
        [-0.6430,  0.7113,  0.4000, -1.2039, -0.4198, -1.1929]])


### **Bài 6. Phép toán và nhân ma trận**
Thực hiện các phép toán phần tử và phép nhân ma trận bằng toán tử @.

In [8]:
import torch
a = torch.randn(2, 3)
b = torch.randn(2, 3)
print("a:")
print(a)
print("b:")
print(b)
print("\na + b:")
print(a + b)
print("\na / b:")
print(a / b)
print("\na^2:")
print(a ** 2)
print("\nMatrix multiplication:")
print(a @ b.T)

a:
tensor([[-0.9351,  0.2138, -1.2842],
        [-0.6917, -0.5359,  0.3355]])
b:
tensor([[ 0.2469,  0.0324,  0.4057],
        [ 1.6181,  0.3932, -0.2148]])

a + b:
tensor([[-0.6881,  0.2462, -0.8785],
        [ 0.9264, -0.1428,  0.1207]])

a / b:
tensor([[-3.7867,  6.5926, -3.1655],
        [-0.4275, -1.3632, -1.5622]])

a^2:
tensor([[0.8743, 0.0457, 1.6492],
        [0.4784, 0.2872, 0.1126]])

Matrix multiplication:
tensor([[-0.7450, -1.1532],
        [-0.0521, -1.4020]])


a * b: Phép nhân từng phần tử. Các phần tử ở vị trí tương ứng của hai Tensor sẽ nhân với nhau.

a @ b.T: Phép nhân ma trận giữa ma trận a và ma trận chuyển vị của b.

Shape của a @ b.T: (2, 2). Lấy ma trận (2, 3) nhân với ma trận (3, 2) sẽ tạo ra ma trận (2, 2).

### **Bài 7. Các hàm thống kê trên Tensor**
Tính trung bình, độ lệch chuẩn và tổng tích lũy theo từng cột.

In [9]:
import torch
a = torch.randn(3, 4)
print(a)
print("\nMean:")
print(torch.mean(a, dim=0))
print("\nStandard deviation:")
print(torch.std(a, dim=0))
print("\nCumulative sum:")
print(torch.cumsum(a, dim=0))

tensor([[ 1.2651, -0.3178, -0.4279, -0.5007],
        [-0.1307,  0.9796,  0.5960, -1.4723],
        [-0.4749,  1.6640,  0.0110,  0.1447]])

Mean:
tensor([ 0.2199,  0.7752,  0.0597, -0.6094])

Standard deviation:
tensor([0.9214, 1.0066, 0.5137, 0.8140])

Cumulative sum:
tensor([[ 1.2651, -0.3178, -0.4279, -0.5007],
        [ 1.1344,  0.6618,  0.1681, -1.9730],
        [ 0.6596,  2.3257,  0.1790, -1.8283]])


In [10]:
import torch
a = torch.randn(3, 4)
print(a)
print("\nMean:")
print(torch.mean(a, dim=1))
print("\nStandard deviation:")
print(torch.std(a, dim=1))
print("\nCumulative sum:")
print(torch.cumsum(a, dim=1))

tensor([[ 1.2818, -1.5952, -1.0648,  0.1055],
        [ 1.9739,  1.0703, -1.7379,  0.1841],
        [-0.1563, -0.6681,  1.2135,  0.7924]])

Mean:
tensor([-0.3182,  0.3726,  0.2954])

Standard deviation:
tensor([1.2816, 1.5854, 0.8606])

Cumulative sum:
tensor([[ 1.2818, -0.3134, -1.3782, -1.2727],
        [ 1.9739,  3.0442,  1.3063,  1.4904],
        [-0.1563, -0.8243,  0.3892,  1.1816]])


dim=0: Tính toán theo từng cột (chiều dọc). mean và std trả về 4 giá trị. cumsum cộng dồn từ trên xuống dưới.

dim=1: Tính toán theo từng hàng (chiều ngang). mean và std trả về 3 giá trị. cumsum cộng dồn từ trái sang phải.

### **Bài 8. Autograd cơ bản**
Tạo biến có requires_grad=True và dùng backward() để tính đạo hàm tự động.

In [11]:
import torch
x = torch.tensor(
3.6,
requires_grad=True
)
y = x * x
y.backward()
print("x =", x)
print("y =", y)
print("x.grad =", x.grad)

x = tensor(3.6000, requires_grad=True)
y = tensor(12.9600, grad_fn=<MulBackward0>)
x.grad = tensor(7.2000)


Tính thủ công:
- Hàm số: $y = x^2$
- Đạo hàm: $y' = 2x$
- Tại $x = 3.6$, giá trị đạo hàm là $y' = 2 \times 3.6 = 7.2$.
- **Nhận xét:** Kết quả tính thủ công khớp hoàn toàn với giá trị x.grad.

Code chạy thử cho hàm $y = 3x^2 + 2x + 1$:

In [12]:
import torch
x = torch.tensor(3.6, requires_grad=True)
y = 3 * x**2 + 2 * x + 1
y.backward()
print("x =", x)
print("y =", y)
print("x.grad =", x.grad)

x = tensor(3.6000, requires_grad=True)
y = tensor(47.0800, grad_fn=<AddBackward0>)
x.grad = tensor(23.6000)


Hàm số: $y = 3x^2 + 2x + 1$

Đạo hàm: $y' = 6x + 2$

Tại $x = 3.6$:
- Giá trị $y = 3(3.6)^2 + 2(3.6) + 1 = 47.08$ (khớp với kết quả).
- Giá trị đạo hàm $y' = 6(3.6) + 2 = 23.6$ (khớp với x.grad).

### **Bài 9. Gradient Descent tìm hệ số đa thức**
Sinh dữ liệu từ đa thức y = x2 + 2x + 3, sau đó dùng NAdam và autograd để tìm lại ba hệ số.

In [13]:
import numpy as np
import torch

np.random.seed(42)
torch.manual_seed(42)

polynomial = np.poly1d([1, 2, 3])
N = 20
X = np.random.randn(N, 1) * 5
Y = polynomial(X)
XX = np.hstack([
X * X,
X,
np.ones_like(X)
])
w = torch.randn(3, 1, requires_grad=True)
x = torch.tensor(XX, dtype=torch.float32)
y = torch.tensor(Y, dtype=torch.float32)
optimizer = torch.optim.NAdam([w], lr=0.01)
print("Initial coefficients:")
print(w)
for epoch in range(3000):
    optimizer.zero_grad()
    y_pred = x @ w
    mse = torch.mean(torch.square(y - y_pred))
    mse.backward()
    optimizer.step()
print("Learned coefficients:")
print(w)
print("Final MSE:", mse.item())

Initial coefficients:
tensor([[0.3367],
        [0.1288],
        [0.2345]], requires_grad=True)


Learned coefficients:
tensor([[1.0000],
        [2.0000],
        [3.0000]], requires_grad=True)
Final MSE: 6.618250947276749e-10


So sánh hệ số:
- Hệ số học được [1.0000, 2.0000, 3.0000] gần như trùng khớp hoàn toàn với hệ số gốc [1, 2, 3] (MSE cuối ~6.6e-10). Việc đặt seed cố định và tăng số epoch lên 3000 giúp mô hình hội tụ ổn định hơn.

Giải thích các hàm:
- optimizer.zero_grad(): Xóa gradient của bước trước về 0 để tránh tình trạng cộng dồn sai lệch.
- mse.backward(): Tính đạo hàm của hàm mất mát theo tham số cần tối ưu (w).
- optimizer.step(): Cập nhật lại giá trị tham số dựa trên gradient vừa tính.

### **Bài 10. Giải hệ phương trình bằng Autograd**
Tối thiểu hóa tổng sai số bình phương để tìm A, B, C, D thỏa bốn phương trình.

In [14]:
import random
import torch
A = torch.tensor(random.random(), requires_grad=True)
B = torch.tensor(random.random(), requires_grad=True)
C = torch.tensor(random.random(), requires_grad=True)
D = torch.tensor(random.random(), requires_grad=True)
EPOCHS = 2000
optimizer = torch.optim.NAdam([A, B, C, D], lr=0.01)
for _ in range(EPOCHS):
    y1 = A + B - 9
    y2 = C - D - 1
    y3 = A + C - 8
    y4 = B - D - 2
    
    sqerr = y1*y1 + y2*y2 + y3*y3 + y4*y4

    optimizer.zero_grad()
    sqerr.backward()
    optimizer.step()
print("A =", A)
print("B =", B)
print("C =", C)
print("D =", D)

A = tensor(4.5880, requires_grad=True)
B = tensor(4.4120, requires_grad=True)
C = tensor(3.4120, requires_grad=True)
D = tensor(2.4120, requires_grad=True)


Kiểm tra nghiệm:

Các giá trị hoàn toàn thỏa mãn hệ phương trình (với sai số cực nhỏ):
- $A + B = 8.9999 \approx 9$
- $C - D = 1.0001 \approx 1$
- $A + C = 8$
- $B - D = 2$

Lý do hệ có nhiều nghiệm:

- Hệ bị phụ thuộc tuyến tính: Phương trình 4 là hệ quả của 3 phương trình đầu: $(A+B) - (A+C) + (C-D) = B - D = 2$.
- Có 4 ẩn nhưng chỉ 3 phương trình độc lập dẫn đến vô số nghiệm. Thuật toán chỉ tìm ra một nghiệm ngẫu nhiên phụ thuộc vào giá trị khởi tạo ban đầu.

### **Bài 11. Viết chương trình PyTorch để thực hiện các yêu cầu sau:**
- Tạo Tensor X có 24 phần tử và reshape lần lượt thành 4 × 6, 2 × 12 và 2 × 3 × 4.
- Tạo hai Tensor A, B kích thước 3 × 4; tính A+B, A*B, A/B, trung bình theo cột và độ lệch chuẩn theo cột.
- Cho y = 2x2 + 5x + 3. Dùng autograd tính dy/dx tại x = 2 và kiểm tra bằng phép tính thủ công.
- Thay đa thức trong Bài 9 bằng y = 2x2 - 3x + 5; huấn luyện và cho biết các hệ số tìm được.

In [15]:
import torch
import numpy as np

torch.manual_seed(0)
np.random.seed(0)

# Yêu cầu 1: Khởi tạo và Reshape
X = torch.arange(24)
X_4x6 = X.reshape(4, 6)
X_2x12 = X.reshape(2, 12)
X_2x3x4 = X.reshape(2, 3, 4)

print(X_4x6.shape)
print(X_2x12.shape)
print(X_2x3x4.shape)

# Yêu cầu 2: Phép toán và Thống kê
A = torch.randn(3, 4)
B = torch.randn(3, 4)

print(A + B)
print(A * B)
print(A / B)
print(torch.mean(A, dim=0))
print(torch.std(A, dim=0))

# Yêu cầu 3: Autograd
x = torch.tensor(2.0, requires_grad=True)
y = 2 * x**2 + 5 * x + 3
y.backward()
print(x.grad.item())

# Yêu cầu 4: Gradient Descent tìm hệ số
N = 100
X_np = np.random.randn(N, 1) * 5
Y_np = 2 * X_np**2 - 3 * X_np + 5

XX = np.hstack([X_np**2, X_np, np.ones_like(X_np)])
w = torch.randn(3, 1, requires_grad=True)

X_tensor = torch.tensor(XX, dtype=torch.float32)
Y_tensor = torch.tensor(Y_np, dtype=torch.float32)

optimizer = torch.optim.NAdam([w], lr=0.01)

for epoch in range(5000):
    optimizer.zero_grad()
    Y_pred = X_tensor @ w
    mse = torch.mean((Y_tensor - Y_pred)**2)
    mse.backward()
    optimizer.step()

print("\nHệ số học được (so với [2, -3, 5]):\n", w.detach().numpy())
print("Final MSE:", mse.item())

torch.Size([4, 6])
torch.Size([2, 12])
torch.Size([2, 3, 4])
tensor([[ 0.6843,  0.8072, -3.2500,  0.6911],
        [-1.6508, -1.0255, -0.4886, -0.6711],
        [-0.3489,  1.0532,  0.3432,  0.9569]])
tensor([[-1.3201, -0.3229,  2.3339,  0.0697],
        [ 0.6142, -0.5218, -0.3598, -1.2647],
        [-0.2664, -0.5875, -0.5607,  0.1411]])
tensor([[-1.7988, -0.2666,  2.0340,  4.6326],
        [ 1.9150, -3.7484, -0.4522, -0.5553],
        [-1.9419, -0.2769, -0.6348,  0.2349]])
tensor([-0.0876, -0.6985, -0.7907,  0.5295])
tensor([1.4222, 0.6088, 1.3020, 0.3297])
13.0



Hệ số học được (so với [2, -3, 5]):
 [[ 2.000019 ]
 [-3.0000048]
 [ 4.998782 ]]
Final MSE: 9.284929660680064e-07


Kiểm tra thủ công (Yêu cầu 3): Đạo hàm của $y = 2x^2 + 5x + 3$ là $y' = 4x + 5$. Thay $x = 2$, ta được kết quả là 13.0, khớp chính xác với kết quả từ x.grad.

Hệ số học được (Yêu cầu 4): Nhờ việc sử dụng 5000 epoch, sai số (MSE) đã giảm xuống mức cực kỳ nhỏ ($9.28 \times 10^{-7}$). 

Các hệ số học được là [2.000019, -3.0000048, 4.998782], gần như tuyệt đối chính xác so với hệ số gốc của đa thức là [2, -3, 5].